### PR TITLE
Add Functionality To Limit Queued Jobs

### PR DESCRIPTION
**Pull request by hantaowang**
The `limit_queued_job` utility checks how many `IN_PROGESS` and `QUEUED` jobs a source has, and then automatically fails the job if adding it to the queue brings the source over `MAX_JOBS_ALLOWED`.

### CODE REVIEW

**Comment by hantaowang** [comment]
Is it possible for this project to be donated to [CS 162](http://github.com/Berkeley-CS162) and/or [ASUC OCTO](http://github.com/asuc-octo) (who runs Berkeleytime.com) for continued open source development? @rogerhub @aykamko

**Comment by aykamko** [comment]
Fine with me! @rogerhub thoughts?

File: ob2/util/job_limiter.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+MAX_JOBS_ALLOWED = 1
+
+def limit_queued_job(repo_name, build_name, limit):
+    try:
+        with DbCursor() as c:
+            c.execute('''SELECT count(*) FROM builds
+                         WHERE source = ? AND (status = ? OR status = ?)''',
+                      [repo_name, QUEUED, IN_PROGRESS])
+            count = int(c.fetchone()[0])
+            if count > limit:
```
Comment (rogerhub) [inline ob2/util/job_limiter.py]: I think this ought to be in ob2.database.helpers.create_build. If there are too many queued, just raise an exception and don't create a new record in the builds table?

File: ob2/util/job_limiter.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+                message = "Cannot have more than {} builds in progress or queued.".format(limit)
+                c.execute('''UPDATE builds SET status = ?, updated = ?, log = ?
+                             WHERE build_name = ?''', [FAILED, now_str(), message, build_name])
+                return False
```
Comment (rogerhub) [inline ob2/util/job_limiter.py]: It's a bit unintuitive that a function named 'limit_queued_job' would return a boolean.

File: ob2/util/job_limiter.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+            return True
+    except Exception as e:
```
Comment (rogerhub) [inline ob2/util/job_limiter.py]: What is this intended to catch?

File: ob2/web/blueprints/dashboard/__init__.py [hunk 4]
```diff
@@ -29,6 +29,7 @@
 from ob2.util.security import require_csrf_token
 from ob2.util.time import now_compare, slip_units_now
 from ob2.util.validation import fail_validation, ValidationError, redirect_with_error
+from ob2.util.job_limiter import limit_queued_job, MAX_JOBS_ALLOWED
```
Comment (rogerhub) [inline ob2/web/blueprints/dashboard/__init__.py]: We should make MAX_JOBS_ALLOWED configurable. (see config/config.yaml)

Also, shouldn't this default to the current behavior? (unlimited)

**Review by rogerhub — CHANGES_REQUESTED** [review]
@hantaowang I added you to the github organization, so do what you want. I'm not aware of any users other than CS162.

File: ob2/util/job_limiter.py (hunk 1 above)
Comment (hantaowang) [reply ob2/util/job_limiter.py]: I would rather have a build created and then failed, so the user can see / be redirected to a failed page with a reason as to why.

Comment (rogerhub) [reply ob2/util/job_limiter.py]: That seems like a lazy way to accomplish the web UI + email notification for quota overages. Admission control should be separate from failures in the grading itself.

As an alternative, why not cancel old builds when new builds are scheduled? (with a specified concurrency limit) That's usually the intended behavior anyway.

Also, in the current implementation, wouldn't this silently cancel builds without sending any email notification? Is that intentional?

Comment (hantaowang) [reply ob2/util/job_limiter.py]: Thats much more work than I have time to do - we just had an issue with groups spamming the autograder for to pass sync tests. I'm not actually sure that failing old builds is the right approach. Maybe we can schedule additional jobs by the same source in a separate low priority queue such that there's only ever MAX_JOBS_ALLOWED jobs in the normal queue + workers.

I'll pass these to TAs next semester or whoever I get to work on this.

**Pull request closed without merging**